packages/byods/src/data-source-service/index.ts

```diff
@@ -0,0 +1,160 @@
+   * @returns {Promise<void>}
+   */
+
+  private async startAutoRefresh(
+    dataSourceId: string,
+    tokenLifetimeMinutes: number,
+    nonceGenerator: () => string = uuidv4 // Defaults to uuidv4 if nonceGenerator is not provided.
+  ): Promise<void> {
+    try {
+      // Call the get method to fetch the dataSource details & then call the update method to update the token!
+      const payloadForDataSourceUpdateMethod: any = {};
+      await this.dataSourceClient
+        .get(dataSourceId)
+        .then((response: any) => {
+          const jwsToken: any = response.jwsToken;
+          const jwsTokenPayload: any = this.decodeJWSTokenAndGetPayload(jwsToken);
+          payloadForDataSourceUpdateMethod.schemaId = response.schemaId;
+          payloadForDataSourceUpdateMethod.url = jwsTokenPayload.payload.com.cisco.datasource.url;
+          payloadForDataSourceUpdateMethod.subject = jwsTokenPayload.payload.sub;
+          payloadForDataSourceUpdateMethod.audience = jwsTokenPayload.payload.aud;
+          payloadForDataSourceUpdateMethod.nonce = nonceGenerator(); // Use the provided nonce generator
+          payloadForDataSourceUpdateMethod.tokenLifetimeMinutes = tokenLifetimeMinutes;
+        })
+        .catch((error: any) => {
+          console.log('error while calling the get method to update the dataSource', error);
+        });
+      setInterval(async () => {
+        await this.dataSourceClient
+          .update(dataSourceId, payloadForDataSourceUpdateMethod)
+          .then((response) => {
+            console.log('dataSource has been updated successfully!', response);
+          })
+          .catch((error) => {
+            console.log('error while updating the dataSource', error);
+          });
+        console.log('dataSource has been refreshed successfully');
+      }, tokenLifetimeMinutes * 60 * 1000); // Converts minutes to milliseconds.
+    } catch (error) {
+      console.log('Got error while starting auto refreshing dataSource token', error);
+    }
+  }
+
+  /**
+   * This method will stop auto refreshing the DataSource token.
+   * @returns {void}
+   * @example
+   * const dataSourceService = new DataSourceService(dataSourceClient); Create an instance of DataSourceService class.
+   * dataSourceService.stopAutoRefresh(); call the stopAutoRefresh method to stop refreshing the dataSource token.
+   */
+
+  public stopAutoRefresh(): void {
+    if (this.timer) {
+      clearInterval(this.timer);
+      console.log('timer has been cleared successfully.');
+    } else {
+      console.info('timer has not started yet!');
+    }
+  }
+
+  /**
+   * This Private method decodes the JWS token that accepts the parameter which is a token.
+   * @param {string} token A JWS token
+   * @returns {Promise<JWTPayload>} This method return a promise which eventually resolved to decoded payload.
+   */
+
+  private async decodeJWSTokenAndGetPayload(token: string): Promise<{payload: JWTPayload} | null> {
```

Comment:
this function not required, decodeJWT can be called directly as its not async

Reply:
**_Addressed!_**